compiler/rustc_middle/src/ty/context.rs

```diff
@@ -1954,33 +1954,108 @@ impl<'tcx> TyCtxt<'tcx> {
         if pred.kind() != binder { self.mk_predicate(binder) } else { pred }
     }
 
-    #[inline(always)]
-    pub(crate) fn check_and_mk_args(
+    pub fn check_args_compatible(self, def_id: DefId, args: &'tcx [ty::GenericArg<'tcx>]) -> bool {
+        self.check_args_compatible_inner(def_id, args, false)
+    }
+
+    fn check_args_compatible_inner(
         self,
-        _def_id: DefId,
-        args: impl IntoIterator<Item: Into<GenericArg<'tcx>>>,
-    ) -> GenericArgsRef<'tcx> {
-        let args = args.into_iter().map(Into::into);
-        #[cfg(debug_assertions)]
+        def_id: DefId,
+        args: &'tcx [ty::GenericArg<'tcx>],
+        nested: bool,
+    ) -> bool {
+        let generics = self.generics_of(def_id);
+
+        // IATs themselves have a weird arg setup (self + own args), but nested items *in* IATs
+        // (namely: opaques, i.e. ATPITs) do not.
```

Comment:
do we have a test that exercises this or could you post a snippet? haven't wrapped my head around that yet

Reply:
consider an ITIATII (impl trait in associated type in inherent impl):

```
impl<T, U> Struct<T, U> {
  type Assoc = impl Trait;
}
```

The `Assoc` type will have substs `[Struct<T, U>]` but the `impl Trait` will have substs `[T, U]`. So if we recurse on the parent when checking the opaque in `check_args_compatible_inner`, we'll ICE.